File: algorithms/[173]二叉搜索树迭代器/solution.py
# topics = ["树", "栈"]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from __future__ import annotations
from typing import List, Optional, Union


class TreeNode:
    def __init__(
        self,
        x: int = 0,
        left: Optional[TreeNode] = None,
        right: Optional[TreeNode] = None,
    ):
        self.val = x
        self.left = left
        self.right = right


class BSTIterator:
    def __init__(self, root: TreeNode):
        self.iter: Union[TreeNode, None] = root
        self.stack: List[TreeNode] = []

    def next(self) -> int:
        while self.iter:
            self.stack.append(self.iter)
            self.iter = self.iter.left

        self.iter = self.stack.pop()
        val = self.iter.val
        self.iter = self.iter.right

        return val

    def hasNext(self) -> bool:
        return self.iter is not None or len(self.stack) != 0


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: algorithms/[173]二叉搜索树迭代器/solution.ts
// topics = ["树", "栈"]

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

// 同 94.二叉树的中序遍历
export class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class BSTIterator {
  stack: TreeNode[];
  iter: TreeNode | null;
  constructor(root: TreeNode | null) {
    this.stack = [];
    this.iter = root;
  }

  next(): number {
    while (this.iter != null) {
      this.stack.push(this.iter!);
      this.iter = this.iter!.left;
    }
    this.iter = this.stack.pop()!;
    const val = this.iter?.val;
    this.iter = this.iter.right;

    return val;
  }

  hasNext(): boolean {
    return this.iter != null || this.stack.length != 0;
  }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
